### PR TITLE
[FLINK-32558][flink-java] Deprecate all DataSet API

### PR DESCRIPTION
**Pull request by WencongLiu**
## What is the purpose of the change

The entry points for using the DataSet API were deprecated. This PR ensures that all APIs with Public/PublicEvolving/Experimental annotation are also marked as deprecated.

## Brief change log

  - All classes that contain one of those methods are also marked as deprecated with a link to [FLIP-131](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=158866741) .

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / no / don't know)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

File: flink-examples/flink-examples-batch/pom.xml [hunk 1]
```diff
@@ -48,25 +48,6 @@ under the License.
...
-							<goal>compile</goal>
-						</goals>
-						<configuration>
-							<compilerArgument>-Xlint:deprecation</compilerArgument>
-							<failOnWarning>true</failOnWarning>
-						</configuration>
-					</execution>
-				</executions>
-			</plugin>
-
```
Comment (xintongsong) [inline flink-examples/flink-examples-batch/pom.xml:69]: I wonder if we should simply remove this module. The original purpose is to not showing user outdated examples. With DataSet API being deprecated, we don't want any newcomers to use it.

File: flink-core/src/main/java/org/apache/flink/api/common/ExecutionConfig.java [hunk 2]
```diff
@@ -556,7 +556,14 @@ public void setExecutionMode(ExecutionMode executionMode) {
...
      *
      * @return The execution mode for the program.
+     * @deprecated All Flink DataSet APIs are deprecated since Flink 1.18 and will be removed in a
+     *     future Flink major version. You can still build your application in DataSet, but you
+     *     should move to either the DataStream and/or Table API.
+     * @see <a href="https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=158866741">
+     *     FLIP-131: Consolidate the user-facing Dataflow SDKs/APIs (and deprecate the DataSet
+     *     API</a>
      */
+    @Deprecated
```
Comment (xintongsong) [inline flink-core/src/main/java/org/apache/flink/api/common/ExecutionConfig.java:566]: Shouldn't we also deprecated the class `ExecutionMode` and the method `setExecutionMode`?

Comment (xintongsong) [reply flink-core/src/main/java/org/apache/flink/api/common/ExecutionConfig.java:566]: Also, there are some descriptions about the execution mode in the class JavaDoc, which should also be updated.

File: flink-core/src/main/java/org/apache/flink/api/common/ExecutionConfig.java [hunk 3]
```diff
@@ -913,6 +920,18 @@ public LinkedHashSet<Class<?>> getRegisteredPojoTypes() {
...
+     *
+     * @return if the auto type registration is disabled.
+     * @deprecated All Flink DataSet APIs are deprecated since Flink 1.18 and will be removed in a
+     *     future Flink major version. You can still build your application in DataSet, but you
+     *     should move to either the DataStream and/or Table API.
+     * @see <a href="https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=158866741">
+     *     FLIP-131: Consolidate the user-facing Dataflow SDKs/APIs (and deprecate the DataSet
+     *     API</a>
+     */
+    @Deprecated
```
Comment (xintongsong) [inline flink-core/src/main/java/org/apache/flink/api/common/ExecutionConfig.java:934]: We should also deprecated `disableAutoTypeRegistration`, and `AUTO_TYPE_REGISTRATION`.

**Review by xintongsong — COMMENTED** [review]
I find the following classes from `flink-java` that are annotated as `@Public` or `@PublicEvolving`, seem to be only used in DataSet scenarios, but not annotated as `@Deprecated`.
- FunctionAnnotation
- DiscardingOutputFormat
- NumericColumnSummary
- AbstractParameterTool
- MultipleParameterTool
- ParameterTool

File: flink-examples/flink-examples-batch/pom.xml (hunk 1 above)
Comment (xintongsong) [reply flink-examples/flink-examples-batch/pom.xml:69]: And we probably should also remove the DataSet related documentations.

**Comment by WencongLiu** [comment]
> * PENDING
`ParameterTool` is still used by the`DataStreamAllroundTestJobFactory`, `HistoryServer` and the DataStream e2e test classes.
`MultipleParameterTool ` is also used by the DataStream e2e test classes.

So I think these two classes and `AbstractParameterTool` shouldn't be deprecated in 1.18 and  can be moved to other modules in the future？

**Comment by xintongsong** [comment]
I think being used in test cases probably should not block the deprecation of the classes. In this specific case, I don't it's necessary to move them to another module. There are many similar classes that parses the application arguments into a `GlobalJobParameters`. I suspect these non-DataSet test cases using `ParameterTool` simply because they are imitated from DataSet test cases.

File: flink-core/src/main/java/org/apache/flink/api/common/ExecutionConfig.java [hunk 4]
```diff
@@ -544,6 +544,13 @@ public ExecutionConfig setExecutionRetryDelay(long executionRetryDelay) {
...
      * @param executionMode The execution mode to use.
+     * @deprecated The {@link ExecutionMode} is deprecated because it's only used in DataSet APIs.
+     *     All Flink DataSet APIs are deprecated since Flink 1.18 and will be removed in a future
+     *     Flink major version. You can still build your application in DataSet, but you should move
+     *     to either the DataStream and/or Table API.
+     * @see <a href="https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=158866741">
+     *     FLIP-131: Consolidate the user-facing Dataflow SDKs/APIs (and deprecate the DataSet
+     *     API</a>
      */
     public void setExecutionMode(ExecutionMode executionMode) {
```
Comment (xintongsong) [inline flink-core/src/main/java/org/apache/flink/api/common/ExecutionConfig.java:555]: Annotation is missing.

File: flink-examples/flink-examples-batch/src/main/java/org/apache/flink/examples/java/clustering/KMeans.java [hunk 5]
```diff
@@ -77,12 +80,23 @@
...
 public class KMeans {
 
+    private static final Logger LOGGER = LoggerFactory.getLogger(KMeans.class);
+
     public static void main(String[] args) throws Exception {
 
+        LOGGER.warn(
+                "All Flink DataSet APIs are deprecated since Flink 1.18 and will be removed in a future"
+                        + " Flink major version. You can still build your application in DataSet, but you should move to"
+                        + " either the DataStream and/or Table API. This class is retained for testing purposes.");
```
Comment (xintongsong) [inline flink-examples/flink-examples-batch/src/main/java/org/apache/flink/examples/java/clustering/KMeans.java:98]: Can we deduplicate the warning message somewhere?

File: flink-examples/flink-examples-batch/pom.xml (hunk 1 above)
Comment (WencongLiu) [reply flink-examples/flink-examples-batch/pom.xml:69]: Fixed.

File: flink-core/src/main/java/org/apache/flink/api/common/ExecutionConfig.java (hunk 4 above)
Comment (WencongLiu) [reply flink-core/src/main/java/org/apache/flink/api/common/ExecutionConfig.java:556]: Fixed.

File: flink-examples/flink-examples-batch/src/main/java/org/apache/flink/examples/java/clustering/KMeans.java (hunk 5 above)
Comment (WencongLiu) [reply flink-examples/flink-examples-batch/src/main/java/org/apache/flink/examples/java/clustering/KMeans.java]: Fixed.

**Comment by WencongLiu** [comment]
> I think being used in test cases probably should not block the deprecation of the classes. In this specific case, I don't it's necessary to move them to another module. There are many similar classes that parses the application arguments into a `GlobalJobParameters`. I suspect these non-DataSet test cases using `ParameterTool` simply because they are imitated from DataSet test cases.

Deprecating `ParameterTool` related classes and make other classes that parse the arguments in similar way with `ParameterTool` is a reasonable method.  I'll follow the suggestions.

File: flink-core/src/main/java/org/apache/flink/api/common/ExecutionConfig.java (hunk 2 above)
Comment (WencongLiu) [reply flink-core/src/main/java/org/apache/flink/api/common/ExecutionConfig.java:575]: 1. `ExecutionMode` and `setExecutionMode ` are both deprecated.
2. The class JavaDoc of execution mode is updated.

File: flink-core/src/main/java/org/apache/flink/api/common/ExecutionConfig.java (hunk 3 above)
Comment (WencongLiu) [reply flink-core/src/main/java/org/apache/flink/api/common/ExecutionConfig.java:944]: Both `disableAutoTypeRegistration ` and `AUTO_TYPE_REGISTRATION ` are deprecated.

File: flink-core/src/main/java/org/apache/flink/api/common/ExecutionConfig.java (hunk 4 above)
Comment (WencongLiu) [reply flink-core/src/main/java/org/apache/flink/api/common/ExecutionConfig.java:556]: The annotation is added.

File: flink-examples/flink-examples-batch/src/main/java/org/apache/flink/examples/java/clustering/KMeans.java (hunk 5 above)
Comment (WencongLiu) [reply flink-examples/flink-examples-batch/src/main/java/org/apache/flink/examples/java/clustering/KMeans.java]: I have added a static String instant DATASET_DEPRECATION_INFO to contain the message.

**Comment by WencongLiu** [comment]
@flinkbot run azure

File: flink-examples/flink-examples-batch/pom.xml (hunk 1 above)
Comment (pegasas) [reply flink-examples/flink-examples-batch/pom.xml:69]: Thanks @xintongsong , @WencongLiu ,
I make a PR for tracking: https://github.com/apache/flink/pull/23135
will request a review after going through it with myself.

**Comment by WencongLiu** [comment]
The CI has passed. Thanks for the careful review. @xintongsong

**Review by xintongsong — APPROVED** [review]
LGTM, merging.

**Pull request closed without merging**